src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -38,89 +41,420 @@ public static ControlFlowGraph Create(IBlockOperation body)
+                            }
+                        }
+
+                        if (region.HasRegions)
+                        {
+                            bool blocksChanged = false;
+
+                            for (int i = region.Regions.Count - 1; i >= 0; i--)
+                            {
+                                RegionBuilder subRegion = region.Regions[i];
+                            
+                                if (subRegion.Kind == ControlFlowGraph.RegionKind.Locals && !subRegion.HasRegions && subRegion.Blocks?.Count == 1)
+                                {
+                                    BasicBlock block = subRegion.Blocks[0];
+
+                                    if (block.Statements.IsEmpty && block.InternalConditional.Condition == null)
+                                    {
+                                        // This sub-region has no executable code, merge block into the parent and drop the sub-region
+                                        region.Add(block);
+                                        blocksChanged = true;
+                                        regionMap[block] = region;
+                                        subRegion.Free();
+                                        region.Regions.RemoveAt(i);
+                                        result = true;
+                                    }
+                                }
+                            }
+
+                            if (blocksChanged)
+                            {
+                                region.Blocks.Sort((x, y) => x.Ordinal.CompareTo(y.Ordinal));
+                            }
+                        }
+
+                        break;
+
+                    case ControlFlowGraph.RegionKind.TryAndCatch:
+                    case ControlFlowGraph.RegionKind.TryAndFinally:
+                    case ControlFlowGraph.RegionKind.FilterAndHandler:
+                        break;
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(region.Kind);
+                }
+
+                return result;
+            }
         }
 
         /// <summary>
         /// Do a pass to eliminate blocks without statements that can be merged with predecessor(s).
+        /// Returns true if any blocks were eliminated
         /// </summary>
-        private static void Pack(ArrayBuilder<BasicBlock> blocks)
+        private static bool PackBlocks(ArrayBuilder<BasicBlock> blocks, PooledDictionary<BasicBlock, RegionBuilder> regionMap)
         {
+            ArrayBuilder<RegionBuilder> fromCurrent = null;
+            ArrayBuilder<RegionBuilder> fromDestination = null;
+            ArrayBuilder<RegionBuilder> fromPredecessor = null;
+
+            bool anyRemoved = false;
+
             int count = blocks.Count - 1;
             for (int i = 1; i < count; i++)
             {
                 BasicBlock block = blocks[i];
-                Debug.Assert(block.Next != null);
```

Comment:
Can we add an assertion after the graph has been calculated for ensuring that the only blocks with null Nexts are the last one and the last blocks of the catches/finalies? If you already have this, I missed it.

Reply:
> Can we add an assertion after the graph has been calculated for ensuring that the only blocks with null Nexts are the last one and the last blocks of the catches/finalies? If you already have this, I missed it.

I guess I can add an assert that "null" branches are either errors,throws, or SEH dispatch. The rest is easier to cover through tests.

---
In reply to: [167660455](https://github.com/dotnet/roslyn/pull/24770#discussion_r167660455) [](ancestors = 167660455)